src/Illuminate/Support/ReadableHigherOrderCollectionProxy.php

```diff
@@ -0,0 +1,85 @@
+<?php
+
+namespace Illuminate\Support;
+
+class ReadableHigherOrderCollectionProxy
+{
+    /**
+     * The collection being operated on.
+     *
+     * @var \Illuminate\Support\Collection
+     */
+    protected $collection;
+
+    /**
+     * The method being proxied.
+     *
+     * @var string
+     */
+    protected $method;
+
+    /**
+     * Create a new proxy instance.
+     *
+     * @param  \Illuminate\Support\Collection  $collection
+     * @param  string  $method
+     * @return void
+     */
+    public function __construct(Collection $collection, $method)
+    {
+        $this->method = $method;
+        $this->collection = $collection;
+    }
+
+    public static function perform(Collection $collection, $method)
+    {
+        $action = 'filter';
+
+        foreach ($collection->getProxies() as $proxy) {
+            if (starts_with($method, $proxy)) {
+                $action = $proxy;
+            }
+        }
+
+        $method = str_replace(Str::cartesian($collection->getProxies(), ['On', 'By', 'Then']), '', $method);
```

Comment:
I had to add a getProxies method here to obtain the list of proxies. Alternatively, the list could be made public.